applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_symmetric.h

```diff
@@ -0,0 +1,524 @@
+    typedef std::vector<NodeType::Pointer>::iterator NodeIterator;
+    typedef std::vector<double>::iterator DoubleVectorIterator;
+    typedef ModelPart::ConditionsContainerType ConditionsArrayType;
+    typedef array_1d<double,3> array_3d;
+
+    // Type definitions for linear algebra including sparse systems
+    typedef UblasSpace<double, CompressedMatrix, Vector> SparseSpaceType;
+    typedef SparseSpaceType::MatrixType SparseMatrixType;
+    typedef SparseSpaceType::VectorType VectorType;
+
+    // Type definitions for tree-search
+    typedef Bucket< 3, NodeType, NodeVector, NodeTypePointer, NodeIterator, DoubleVectorIterator > BucketType;
+    typedef Tree< KDTreePartition<BucketType> > KDTree;
+
+    /// Pointer definition of MapperVertexMorphingSymmetric
+    KRATOS_CLASS_POINTER_DEFINITION(MapperVertexMorphingSymmetric);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    MapperVertexMorphingSymmetric( ModelPart& rOriginModelPart, ModelPart& rDestinationModelPart, Parameters MapperSettings )
+        : mrOriginModelPart(rOriginModelPart),
+          mrDestinationModelPart(rDestinationModelPart),
+          mMapperSettings(MapperSettings)
+    {
+    }
+
+    /// Destructor.
+    virtual ~MapperVertexMorphingSymmetric()
+    {
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    // --------------------------------------------------------------------------
+    void Initialize() override
+    {
+        BuiltinTimer timer;
+        KRATOS_INFO("ShapeOpt") << "Starting initialization of mapper..." << std::endl;
+
+        CreateFilterFunction();
+
+        mIsMappingInitialized = true;
+
+        Update();
+
+        KRATOS_INFO("ShapeOpt") << "Finished initialization of mapper in " << timer.ElapsedSeconds() << " s." << std::endl;
+    }
+
+    // --------------------------------------------------------------------------
+    void Map( const Variable<array_3d> &rOriginVariable, const Variable<array_3d> &rDestinationVariable) override
+    {
+        if (mIsMappingInitialized == false)
+            Initialize();
+
+        BuiltinTimer timer;
+        KRATOS_INFO("") << std::endl;
+        KRATOS_INFO("ShapeOpt") << "Starting mapping of " << rOriginVariable.Name() << "..." << std::endl;
+
+        Vector values_origin(mrOriginModelPart.Nodes().size()*3);
+        Vector values_destination(mrDestinationModelPart.Nodes().size()*3);
+
+        // Prepare vectors for mapping
+        values_origin.clear();
+        values_destination.clear();
+
+        for(auto& node_i : mrOriginModelPart.Nodes())
+        {
+            const int i = node_i.GetValue(MAPPING_ID);
+            const array_3d& r_nodal_variable = node_i.FastGetSolutionStepValue(rOriginVariable);
+            values_origin[i*3+0] = r_nodal_variable[0];
+            values_origin[i*3+1] = r_nodal_variable[1];
+            values_origin[i*3+2] = r_nodal_variable[2];
+        }
```

Comment:
Are these `MAPPING_ID`s unique for each node? Can we use `IndexPartition` here?

Reply:
Yes its unique for each node, what is IndexPartition?

Reply:
```suggestion
        block_for_each(mrOriginModelPart.Nodes(), [&](const ModelPart::NodeType& rNode) {
            const int i = rNode.GetValue(MAPPING_ID);
            const array_3d& r_nodal_variable = rNode.FastGetSolutionStepValue(rOriginVariable);
            values_origin[i*3+0] = r_nodal_variable[0];
            values_origin[i*3+1] = r_nodal_variable[1];
            values_origin[i*3+2] = r_nodal_variable[2];                
        });
```

Sorry you don't need `IndexPartition`, but above should also work ryt?

https://github.com/KratosMultiphysics/Kratos/blob/869f0eaa2fd5c93837af8c98451e4adf7a4c579f/kratos/tests/cpp_tests/utilities/test_parallel_utilities.cpp#L121-L135